### ISSUE TITLE
Substantially worse results via WebUI vs using transformers directly with same params

### ISSUE THREAD

**Issue by my-other-github-account** [oobabooga/text-generation-webui#2755]
### Describe the bug

I am getting significantly worse results via ooba vs using transformers directly, given otherwise same set of parameters - i.e. I am looking at WizardCoder15B, and get approx 20% worse scores over 164 problems via WebUI vs transformers lib. I am pretty sure I have the paramss set the same.

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

For example, for the following prompt, temperature .2, topp .95, topk 50, WizardLM/WizardCoder-15B-V1.0 model, with results more or less consistent across seeds:

```
Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Create a Python script for this problem:

def by_length(arr):
	"""
	Given an array of integers, sort the integers that are between 1 and 9 inclusive,
	reverse the resulting array, and then replace each digit by its corresponding name from
	"One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine".

	For example:
	  arr = [2, 1, 1, 4, 5, 8, 2, 3]   
			-> sort arr -> [1, 1, 2, 2, 3, 4, 5, 8] 
			-> reverse arr -> [8, 5, 4, 3, 2, 2, 1, 1]
	  return ["Eight", "Five", "Four", "Three", "Two", "Two", "One", "One"]
	
	  If the array is empty, return an empty array:
	  arr = []
	  return []
	
	  If the array has any strange number ignore it:
	  arr = [1, -1, 55] 
			-> sort arr -> [-1, 1, 55]
			-> reverse arr -> [55, 1, -1]
	  return = ['One']
	"""


### Response:
```

Transformers gives me:

```
### Response:Here's the Python script for the problem:

```python
def by_length(arr):
    """
    Given an array of integers, sort the integers that are between 1 and 9 inclusive,
    reverse the resulting array, and then replace each digit by its corresponding name from
    "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine".

    For example:
      arr = [2, 1, 1, 4, 5, 8, 2, 3]   
            -> sort arr -> [1, 1, 2, 2, 3, 4, 5, 8] 
            -> reverse arr -> [8, 5, 4, 3, 2, 2, 1, 1]
      return ["Eight", "Five", "Four", "Three", "Two", "Two", "One", "One"]

      If the array is empty, return an empty array:
      arr = []
      return []

      If the array has any strange number ignore it:
      arr = [1, -1, 55] 
            -> sort arr -> [-1, 1, 55]
            -> reverse arr -> [55, 1, -1]
      return = ['One']
    """

    # Create a dictionary to map digits to names
    digit_to_name = {
        1: "One",
        2: "Two",
        3: "Three",
        4: "Four",
        5: "Five",
        6: "Six",
        7: "Seven",
        8: "Eight",
        9: "Nine"
    }

    # Filter the array to include only digits between 1 and 9
    filtered_arr = [num for num in arr if num >= 1 and num <= 9]

    # Sort the filtered array in reverse order
    sorted_arr = sorted(filtered_arr, reverse=True)

    # Replace each digit in the sorted array with its corresponding name
    result = [digit_to_name[num] for num in sorted_arr]

    return result
```

But WebUI gives me:

```
### Response:## **Solution**

We can solve this problem by following the below steps:

1. Sort the given array in ascending order.
2. Reverse the sorted array.
3. Iterate through the reversed array and check if the current element is a positive integer.
4. If the current element is a positive integer, then add the corresponding string to the result array.
5. If the current element is not a positive integer, ignore it.
6. Return the result array.

Let's implement the above steps in code.<|im_sep|>
```

and promptly cuts off. My param settings are attached, but I am essentially just trying to imitate the WizardCoder HumanEval defaults.

My transformers code is from here: https://github.com/nlpxucan/WizardLM/blob/main/WizardCoder/src/humaneval_gen.py

But in summary:

```
def get_model(
    load_8bit: bool = False,
    base_model: str = "bigcode/starcoder",
):
    assert base_model, (
        "Please specify a --base_model, e.g. --base_model='bigcode/starcoder'"
    )

    tokenizer = AutoTokenizer.from_pretrained(base_model)
    if device == "cuda":
        model = AutoModelForCausalLM.from_pretrained(
            base_model,
            load_in_8bit=load_8bit,
            torch_dtype=torch.float16,
            device_map="auto",
        )
    elif device == "mps":
        model = AutoModelForCausalLM.from_pretrained(
            base_model,
            device_map={"": device},
            torch_dtype=torch.float16,
        )
    model.config.pad_token_id = tokenizer.pad_token_id

    if not load_8bit:
        model.half()  # seems to fix bugs for some users.

    model.eval()
    if torch.__version__ >= "2" and sys.platform != "win32":
        model = torch.compile(model)
    
    return tokenizer, model

### … snip … ###

  generation_config = GenerationConfig(
        pad_token_id=tokenizer.pad_token_id,
        do_sample=True,
        temperature=args.temperature,
        max_length=args.max_len,
        num_return_sequences=args.num_seqs_per_iter,
        eos_token_id=tokenizer.eos_token_id,
        top_p=0.95
    )

encoding = tokenizer(prompt_batch, return_tensors="pt", truncation=True, max_length=args.max_len).to(device)

gen_tokens = model.generate(
                        **encoding,
                        generation_config=generation_config
                    )

gen_seqs = tokenizer.batch_decode(gen_tokens, skip_special_tokens=True)

print(gen_seqs[0])
```

### Screenshot

<img width="1672" alt="Screenshot 2023-06-18 at 1 33 28 PM" src="https://github.com/oobabooga/text-generation-webui/assets/82987814/4ebe7949-7422-41fe-984d-92885781880d">

### Logs

```shell
No errors or warnings
```

### System Info

```shell
RunPod A100 80GB in TheBloke's Runpod Ooba WebUI Template Container.
```

### Comments:

Could it be the:

pad_token_id=tokenizer.pad_token_id

or

eos_token_id=tokenizer.eos_token_id

That I am having trouble with? As I don't see an easy way of doing that same assignment in WebUI. Thank you!

**Comment by oobabooga** [comment]
Can you make a test with `do_sample=False` to see if the results are the same?

**Comment by my-other-github-account** [comment]
With this request (where do_sample is False, seed is 26):

```
request = {
    'prompt': '',
    'max_new_tokens': 2048,
    'do_sample': False,
    'temperature': 0.2,
    'top_p': 0.95,
    'typical_p': 1,
    'epsilon_cutoff': 0,  # In units of 1e-4
    'eta_cutoff': 0,  # In units of 1e-4
    'repetition_penalty': 1,
    'top_k': 50,
    'min_length': 0,
    'no_repeat_ngram_size': 0,
    'num_beams': 1,
    'penalty_alpha': 0,
    'length_penalty': 1,
    'early_stopping': False,
    'mirostat_mode': 0,
    'mirostat_tau': 5,
    'mirostat_eta': 0.1,
    'seed': '',
    'add_bos_token': True,
    'truncation_length': 2048,
    'ban_eos_token': False,
    'skip_special_tokens': True,
    'stopping_strings': []
}

def run(prompt, seed=-1, port = 443, deterministic = True, host='localhost'):
    # Set the prompt and seed for the current request
    request = {}
    request['prompt'] = prompt
    request['seed'] = seed
```

and this prompt:

```
Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Create a Python script for this problem:
from typing import List


def remove_duplicates(numbers: List[int]) -> List[int]:
	""" From a list of integers, remove all elements that occur more than once.
	Keep order of elements left the same as in the input.
	>>> remove_duplicates([1, 2, 3, 2, 4])
	[1, 3, 4]
	"""


### Response:
```

I get a good answer via transformers:

```
Here's the Python script for the problem:\r\n\r\n```python\r\nfrom typing import List\r\n\r\ndef remove_duplicates(numbers: List[int]) -> List[int]:\r\n    \"\"\" From a list of integers, remove all elements that occur more than once.\r\n    Keep order of elements left the same as in the input.\r\n    >>> remove_duplicates([1, 2, 3, 2, 4])\r\n    [1, 3, 4]\r\n    \"\"\"\r\n    unique_numbers = []\r\n    for number in numbers:\r\n        if number not in unique_numbers:\r\n            unique_numbers.append(number)\r\n    return unique_numbers\r\n```\r\n\r\nThe `remove_duplicates` function takes a list of integers as input and returns a new list with all duplicates removed. The function uses a loop to iterate over each element in the input list. For each element, it checks if it already exists in the `unique_numbers` list. If it doesn't, it adds the element to the list. Finally, the function returns the `unique_numbers` list.", "all_code": "Below is an instruction that describes a task. Write a response that appropriately completes the request.\n\n### Instruction:\nCreate a Python script for this problem:\nfrom typing import List\n\n\ndef remove_duplicates(numbers: List[int]) -> List[int]:\n    \"\"\" From a list of integers, remove all elements that occur more than once.\n    Keep order of elements left the same as in the input.\n    >>> remove_duplicates([1, 2, 3, 2, 4])\n    [1, 3, 4]\n    \"\"\"\n\n\n### Response:Here's the Python script for the problem:\r\n\r\n```python\r\nfrom typing import List\r\n\r\ndef remove_duplicates(numbers: List[int]) -> List[int]:\r\n    \"\"\" From a list of integers, remove all elements that occur more than once.\r\n    Keep order of elements left the same as in the input.\r\n    >>> remove_duplicates([1, 2, 3, 2, 4])\r\n    [1, 3, 4]\r\n    \"\"\"\r\n    unique_numbers = []\r\n    for number in numbers:\r\n        if number not in unique_numbers:\r\n            unique_numbers.append(number)\r\n    return unique_numbers\r\n```\r\n\r\nThe `remove_duplicates` function takes a list of integers as input and returns a new list with all duplicates removed. The function uses a loop to iterate over each element in the input list. For each element, it checks if it already exists in the `unique_numbers` list. If it doesn't, it adds the element to the list. Finally, the function returns the `unique_numbers` list.
```

and a bad one via WebUI API that repeats and cuts off:

```
### Response:Here's one possible solution to the problem using a dictionary to keep track of occurrences and then filtering out duplicates from the original list:

```python
from typing import List


def remove_duplicates(numbers: List[int]) -> List[int]:
    """ From a list of integers, remove all elements that occur more than once.
    Keep order of elements left the same as in the input.
    >>> remove_duplicates([1, 2, 3, 2, 4])
    [1, 3, 4]
    """

    # Initialize a dictionary to store counts of each number
    count = {}
    
    # Iterate over numbers and update their counts in the dictionary
    for num in numbers:
        if num not in count:
            count[num] = 0
        count[num] += 1
        
    # Filter out duplicate numbers by only keeping those with count 1
    return filter(lambda x: count[x] == 1

from typing import List


def remove_duplicates(numbers: List[int]) -> List[int]:
    """ From a list of integers, remove all elements that occur more than once.
    Keep order of elements left the same as in the input.
    >>> remove_duplicates([1, 2, 3, 2, 4])
    [1, 3, 4]
    """

    # Initialize a dictionary to store counts of each number
    count = {}
    
    # Iterate over numbers and update their counts in the dictionary
    for num in numbers:
        if num not in count:
            count[num] = 0
        count[num] += 1
        
    # Filter out duplicate numbers by only keeping those with count 1
    return filter(lambda x: count[x] == 1
```

**Comment by my-other-github-account** [comment]
Closed - this was my bad, I had GPT4 rewrite my code and missed the part where it added request = {} - sorry for the trouble!
